src/main/java/com/google/api/gax/bundling/ThresholdBundler.java

```diff
@@ -115,30 +127,42 @@ private Builder() {
   /**
    * Adds an element to the bundler. If the element causes the collection to go past any of the
    * thresholds, the bundle will be made available to consumers.
+   *
+   * @throws FlowControlException
    */
-  public void add(E e) {
+  public void add(E e) throws FlowControlException {
+    final Lock addMethodLock = this.addMethodLock;
     final Lock lock = this.lock;
-    lock.lock();
+    addMethodLock.lock();
     try {
-      boolean signalBundleIsReady = false;
-      if (currentOpenBundle == null) {
-        currentOpenBundle = new Bundle(thresholdPrototypes, maxDelay);
-        currentOpenBundle.start();
-        signalBundleIsReady = true;
-      }
+      flowController.reserve(e);
```

Comment:
I don't think `flowController` needs a lock around it. The FlowController is already thread-safe.